### PR TITLE
Add ports

### PR DESCRIPTION
**Pull request by AZaugg**
This is an implementation similar to #415 , but supports both bitmap:ports and hash:ports

[root@localhost pyroute2]# git log -1 --pretty=oneline
2d77bd24bf19c76a591eddedf787358946de915b Added support for adding type:port to existing ipset
[root@localhost pyroute2]#
[root@localhost pyroute2]# ipset list
Name: bports
Type: bitmap:port
Revision: 1
Header: range 1-65000
Size in memory: 520144
References: 0
Members:

Name: hipport
Type: hash:ip,port
Revision: 2
Header: family inet hashsize 1024 maxelem 65536
Size in memory: 16528
References: 0
Members:
[root@localhost pyroute2]# cat IPSETS.PY
from  ipset import IPSet
a = IPSet()
print "Add bitmap port, single port"
a.add(name='bports', entry='80',etype='port')
print "Add bitmap port, range of ports"
a.add(name='bports', entry='81-90',etype='port')
print "Add Hash:ip,ports single port with default proto"
a.add(name='hipport', entry='192.168.0.1,60',etype='ip,port')
print "Add Hash:ip,ports single port with proto (TCP)"
a.add(name='hipport', entry='192.168.0.2,tcp:61',etype='ip,port')
print "Add Hash:ip,ports single port with proto (UDP)"
a.add(name='hipport', entry='192.168.0.3,udp:62',etype='ip,port')
print "Add Hash:ip,ports port range with proto (UDP)"
a.add(name='hipport', entry='192.168.0.4,udp:70-75',etype='ip,port')
print "Add Hash:ip,ports port range with proto (TCP)"
a.add(name='hipport', entry='192.168.0.5,udp:76-80',etype='ip,port')
[root@localhost pyroute2]#
[root@localhost pyroute2]#
[root@localhost pyroute2]# python IPSETS.PY
Add bitmap port, single port
Add bitmap port, range of ports
Add Hash:ip,ports single port with default proto
Add Hash:ip,ports single port with proto (TCP)
Add Hash:ip,ports single port with proto (UDP)
Add Hash:ip,ports port range with proto (UDP)
Add Hash:ip,ports port range with proto (TCP)
[root@localhost pyroute2]# ipset list
Name: bports
Type: bitmap:port
Revision: 1
Header: range 1-65000
Size in memory: 520144
References: 0
Members:
80
81
82
83
84
85
86
87
88
89
90

Name: hipport
Type: hash:ip,port
Revision: 2
Header: family inet hashsize 1024 maxelem 65536
Size in memory: 16976
References: 0
Members:
192.168.0.4,udp:70
192.168.0.4,udp:74
192.168.0.4,udp:71
192.168.0.5,udp:77
192.168.0.4,udp:72
192.168.0.4,udp:75
192.168.0.5,udp:80
192.168.0.4,udp:73
192.168.0.5,udp:76
192.168.0.5,udp:78
192.168.0.5,udp:79
192.168.0.2,tcp:61
192.168.0.3,udp:62
192.168.0.1,tcp:60

### CODE REVIEW

File: pyroute2/ipset.py [hunk 1]
```diff
@@ -196,7 +203,33 @@ def _entry_to_data_attrs(self, entry, etype, family):
...
+                    attrs += [['IPSET_ATTR_PORT_FROM', int(e)]]
+
+                if etype != 'port':
+                    proto_map = {
+                        '': 1, # default value TCP
+                        'tcp': 1,
+                        'udp': 12,
+                    }
+                    attrs += [['IPSET_ATTR_PROTO', {'attrs': 
+                        [['IPSET_ATTR_NAME', 'x'*proto_map[proto]]]
```
Comment (AZaugg) [inline pyroute2/ipset.py]: I am not happy with this approach, if someone could let me know how to better improve this, ie construct the array using the appropriate flags

Comment (svinota) [reply pyroute2/ipset.py]: Could you please give an example of the target NLA structure? Looks a bit odd, that's true.

Comment (AZaugg) [reply pyroute2/ipset.py]: It should look like:
{'attrs': [('IPSET_ATTR_IP_FROM', 
                    {'attrs': [('IPSET_ATTR_IPADDR_IPV4', '192.168.1.1')]}, 32768), 
                    ('IPSET_ATTR_PORT_FROM', 53, 16384), 
                    ('IPSET_ATTR_PROTO', {'attrs': []})]}, 32768)]}, 32768)],

I am unsure of how to express the protocol(UDP, TCP) as the arguments that need to be passed in.

Comment (AZaugg) [reply pyroute2/ipset.py]: I should have also provided the struct from ipset:
`
struct hash_ipport4_elem {
	__be32 ip;
	__be16 port;
	u8 proto;
	u8 padding;
};
`

Comment (svinota) [reply pyroute2/ipset.py]: I'll try to review the whole module on the weekend. Maybe it's time to refactor it a bit.

And maybe we should plan some IRC-meeting next week regarding this topic.

@ffourcot what do you think?

Comment (AZaugg) [reply pyroute2/ipset.py]: Happy to have a chat, just shoot me through some details. In regards to a refactor, I have already written a wrapper around around the IPSet class, which I think would make a good base and happy to open another PR for it to be reviewed.

I was hoping to get this PR merged in as I need it from a project I am working on. But I am having issues with setting the protocol type of the ipset port added. Not sure how to express the the protocol as a NLA attribute. 

I manage to hack something together, to get it working. When the string passed in has a length of 1, it corresponds to TCP, when string length is 14 it corresponds to UDP. Not actually sure how to determine the correct type that I need to pass in order to do this properly. You able to point me in the right direction?

Comment (svinota) [reply pyroute2/ipset.py]: Looks like there is a mistake in the NLA structure in the code you provided — I looked into the kernel, and `IPSET_ATTR_PROTO` should be `NLA_U8`, not a nested structure. So the kernel interpreted the nested NLA length as a value of `IPSET_ATTR_PROTO`.

Could you please provide me the bash code to do the same? With the ipset tool?

**Comment by ffourcot** [comment]
@svinota ok for a IRC meeting to speak on ipset module. I agree with you, we should probably not use strings here, and use a real python object to manage port ranges.

And yes, IPSET_ATTR_PROTO is a u8. It's a bug that it's marked as recursive in netlink/nfnetlink/ipset.py (many currently unused fields are probably buggy).

I'm working on a proof of concept / cleanup for this feature, I will try to push it somewhere today.

File: pyroute2/ipset.py (hunk 1 above)
Comment (ffourcot) [reply pyroute2/ipset.py]: @svinota @AZaugg : it's a work in progress, but what do you think of https://github.com/ffourcot/pyroute2/commits/portrange ? protocol support is planned but not done. That should be to complicated to add

Comment (AZaugg) [reply pyroute2/ipset.py]: I was thinking of keeping with the core data types(int and str)  and was going to control them via new classes, example:
https://gist.github.com/AZaugg/c5c0648b8e87f09a5255de2e909b4503

`a = Hash('ipsetname, 'ip')
a.create()
a.add('192.168.0.1')`

**Comment by AZaugg** [comment]
Hey @ffourcot, I took a look at your new port classes. At present we can pass in a string.
`a.add(name='hipport', entry='192.168.0.2,tcp:61',etype='ip,port')`

But looking at the ipset hash types there are a couple of different combinations that we can potentially pass in and would break this. We need to extend these new classes to accept the ipaddress/network.

Thoughts?

**Comment by AZaugg** [comment]
New commits work like a charm:

[root@localhost pyroute2]# git log -4 --pretty=oneline
b3704beda652b9252c18376bff2e71ec25e99430 Corrected protocol support for hash ports
2883987f49b3d5b7beaf6593504ad3952c91fc24 Fixed IPSET_ATTR_PROTO and set to u8
41a08cb03df4989f80cccaac55e1009a97dd85f4 Added docstring for _entry_to_data_attrs method
2d77bd24bf19c76a591eddedf787358946de915b Added support for adding type:port to existing ipset
[root@localhost pyroute2]# ipset list
Name: hipport
Type: hash:ip,port
Revision: 2
Header: family inet hashsize 1024 maxelem 65536
Size in memory: 16528
References: 0
Members:

Name: bports
Type: bitmap:port
Revision: 1
Header: range 1-65000
Size in memory: 520144
References: 0
Members:
[root@localhost pyroute2]# cat IPSETS.PY
import sys,os
sys.path.append(os.getcwd())

from  pyroute2.ipset import IPSet
a = IPSet()

print "Add bitmap port, single port type int"
a.add(name='bports', entry='80',etype='port')

print "Add bitmap port, single port type PortEntry"
a.add(name='bports', entry='81-90',etype='port')

print "Add Hash:ip,ports single port with default proto"
a.add(name='hipport', entry='192.168.0.2,90',etype='ip,port')
print "Add Hash:ip,ports single port with proto (TCP)"
a.add(name='hipport', entry='192.168.0.2,tcp:61',etype='ip,port')
print "Add Hash:ip,ports single port with proto (UDP)"
a.add(name='hipport', entry='192.168.0.3,udp:62',etype='ip,port')
print "Add Hash:ip,ports port range with proto (UDP)"
a.add(name='hipport', entry='192.168.0.4,udp:70-75',etype='ip,port')
print "Add Hash:ip,ports port range with proto (TCP)"
a.add(name='hipport', entry='192.168.0.5,tcp:76-80',etype='ip,port')
[root@localhost pyroute2]# python IPSETS.PY
Add bitmap port, single port type int
Add bitmap port, single port type PortEntry
Add Hash:ip,ports single port with default proto
Add Hash:ip,ports single port with proto (TCP)
Add Hash:ip,ports single port with proto (UDP)
Add Hash:ip,ports port range with proto (UDP)
Add Hash:ip,ports port range with proto (TCP)
[root@localhost pyroute2]# ipset list
Name: hipport
Type: hash:ip,port
Revision: 2
Header: family inet hashsize 1024 maxelem 65536
Size in memory: 16976
References: 0
Members:
192.168.0.5,tcp:76
192.168.0.2,tcp:90
192.168.0.5,tcp:79
192.168.0.4,udp:71
192.168.0.5,tcp:77
192.168.0.4,udp:72
192.168.0.4,udp:70
192.168.0.3,udp:62
192.168.0.5,tcp:80
192.168.0.5,tcp:78
192.168.0.4,udp:74
192.168.0.2,tcp:61
192.168.0.4,udp:73
192.168.0.4,udp:75

Name: bports
Type: bitmap:port
Revision: 1
Header: range 1-65000
Size in memory: 520144
References: 0
Members:
80
81
82
83
84
85
86
87
88
89
90

**Comment by AZaugg** [comment]
I dont think that this branch is needed and most of the changes where inserted in #422

**Comment by svinota** [comment]
Do we close this PR then?

**Pull request closed without merging**